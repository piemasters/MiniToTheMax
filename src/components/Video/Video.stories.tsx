import type { Meta, StoryObj } from '@storybook/react-vite';

import { Video } from './Video';

const meta: Meta<typeof Video> = {
  component: Video,
  argTypes: {
    videoId: {
      name: 'videoId',
      description: 'The id part of the source for the video file',
      defaultValue: '',
      control: { type: 'text' },
    },
    title: {
      name: 'title',
      description: 'Title of the video',
      defaultValue: 'Video Title',
      control: { type: 'text' },
    },
    aspectWidth: {
      name: 'aspectWidth',
      description: 'The aspect ratio width of the video',
      defaultValue: 16,
      control: { type: 'number' },
    },
    aspectHeight: {
      name: 'aspectHeight',
      description: 'The aspect ratio height of the video',
      defaultValue: 9,
      control: { type: 'number' },
    },
  },
};

export default meta;
type Story = StoryObj<typeof Video>;

export const Primary: Story = {
  args: {
    videoId: 's1Ck5Xn6420',
    title: 'Grey Knight Librarian',
    aspectWidth: 16,
    aspectHeight: 9,
  },
};
