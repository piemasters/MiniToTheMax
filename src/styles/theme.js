export default {
  colors: {
    text: '#000',
    muted: '#999',
  },
};
