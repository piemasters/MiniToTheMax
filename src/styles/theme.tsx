export interface Theme {
  colors: {
    text: string;
    muted: string;
  };
}

const appTheme: Theme = {
  colors: {
    text: '#000',
    muted: '#999',
  },
};

export default appTheme;
