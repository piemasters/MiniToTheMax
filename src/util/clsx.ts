/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable @typescript-eslint/no-explicit-any */
export type ClassValue =
  | ClassArray
  | ClassDictionary
  | string
  | number
  | null
  | boolean
  | undefined;
type ClassDictionary = Record<string, any>;
type ClassArray = ClassValue[];

function toVal(mix: string | number | boolean | unknown[] | ClassDictionary) {
  let k,
    y,
    str = '';

  if (typeof mix === 'string' || typeof mix === 'number') {
    str += mix;
  } else if (typeof mix === 'object') {
    if (Array.isArray(mix)) {
      const len = mix.length;
      for (k = 0; k < len; k++) {
        if (mix[k]) {
          if ((y = toVal(mix[k]))) {
            str && (str += ' ');
            str += y;
          }
        }
      }
    } else {
      for (y in mix) {
        if (mix[y]) {
          str && (str += ' ');
          str += y;
        }
      }
    }
  }

  return str;
}

export function clsx(...args: ClassValue[]): string {
  let i = 0,
    tmp,
    x,
    str = '';
  const len = args.length;
  for (; i < len; i++) {
    if ((tmp = args[i])) {
      if ((x = toVal(tmp))) {
        str && (str += ' ');
        str += x;
      }
    }
  }
  return str;
}

export default clsx;
