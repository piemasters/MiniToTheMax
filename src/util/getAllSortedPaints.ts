import { PaintDetails } from '../types';
import { textToId } from './textToId';
import { sortPaints } from './sortPaints';
import {
  airPaints,
  basePaints,
  contrastPaints,
  dryPaints,
  layerPaints,
  shadePaints,
  sprayPaints,
  technicalPaints,
} from '../data/paints/citadel';
import { gameColorPaints } from '../data/paints';

export const getAllSortedPaints = (): PaintDetails[] => {
  const allPaints: PaintDetails[] = [
    // citadel
    ...airPaints,
    ...basePaints,
    ...contrastPaints,
    ...dryPaints,
    ...layerPaints,
    ...shadePaints,
    ...sprayPaints,
    ...technicalPaints,
    // vallejo
    ...gameColorPaints,
  ];

  // Remove duplicates using a Set
  const uniquePaints = Array.from(
    new Set(
      allPaints.map(
        (paint) =>
          `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}`
      )
    )
  ).map((uniqueKey) =>
    allPaints.find(
      (paint) =>
        `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}` ===
        uniqueKey
    )
  ) as PaintDetails[];

  // Sort the paints using the sortPaints function
  return uniquePaints.sort(sortPaints);
};
