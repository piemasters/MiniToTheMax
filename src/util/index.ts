export * from './pagination';
export * from './paintFilters';
