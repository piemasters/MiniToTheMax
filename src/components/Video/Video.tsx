import type { FC } from 'react';
import { css, useTheme } from '@emotion/react';

import { Theme } from '../styles/theme';

export interface VideoProps {
  src: string;
  title: string;
  width?: number;
  aspectRatio?: number;
}

export const Video: FC<VideoProps> = ({
  src,
  title,
  width = 714,
  aspectRatio = 16 / 9,
}) => {
  const theme = useTheme() as Theme;

  const videoStyle = css`
    iframe {
      &:hover {
        box-shadow: 6px 6px 60px 5px ${theme.colors.backgroundTransparent};
      }
    }
  `;
  return (
    <div css={videoStyle} data-testid="video">
      <iframe
        src={src}
        title={title}
        allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
        frameBorder="0"
        allowFullScreen
        width={'100%'}
        height={width / aspectRatio}
      />
    </div>
  );
};
export default Video;
