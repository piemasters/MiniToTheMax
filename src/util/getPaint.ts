import {
  CitadelPaintCategories,
  PaintCategory,
  PaintDetails,
  VallejoPaintCategories,
} from '../types';
import { getAllSortedPaints } from './getAllSortedPaints';

const allPaints = getAllSortedPaints();

const paintCategoryPriority = [
  ...CitadelPaintCategories,
  ...VallejoPaintCategories,
] as const;

type category = (typeof paintCategoryPriority)[number];

export const getPaint = (
  name: string,
  category?: PaintCategory
): { paint?: PaintDetails; error?: string; type?: string } => {
  const paintMatches = allPaints.filter((paint) => paint.name === name);

  if (paintMatches.length === 0) {
    return { error: 'Incorrect Paint Name', type: name };
  }

  if (paintMatches.length === 1) {
    return { paint: paintMatches[0] };
  }

  if (category) {
    const categoryMatches = paintMatches.filter(
      (paint) => paint.category === category
    );
    if (categoryMatches.length === 0) {
      return { error: 'Incorrect Paint Category', type: category };
    }
    return { paint: categoryMatches[0] };
  }

  const getCategoryPriority = (category: PaintCategory) => {
    const index = paintCategoryPriority.indexOf(category);
    return index === -1 ? Infinity : index;
  };
  paintMatches.sort(
    (a, b) => getCategoryPriority(a.category) - getCategoryPriority(b.category)
  );

  return { paint: paintMatches[0] };
};
