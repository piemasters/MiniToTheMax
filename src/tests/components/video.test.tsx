import React from 'react';
import Video from '../../components/video';
import { getByTestId } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import theme from '../../styles/theme';
import { renderWithTransitionProvider } from '../util/transition-provider';

describe('Video', () => {
  test('renders correctly', () => {
    const { container } = renderWithTransitionProvider(
      <ThemeProvider theme={theme}>
        <Video
          src="https://www.youtube.com/embed/qcWaykRRDfM"
          title="Grey Knight Stormraven"
        />{' '}
      </ThemeProvider>
    );
    expect(getByTestId(container, 'video')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
