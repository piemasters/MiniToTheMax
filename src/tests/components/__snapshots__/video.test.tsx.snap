// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video renders correctly 1`] = `
<div>
  <div
    data-testid="video"
  >
    <iframe
      allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
      allowfullscreen=""
      frameborder="0"
      height="401.625"
      src="https://www.youtube.com/embed/qcWaykRRDfM"
      title="Grey Knight Stormraven"
      width="714"
    />
  </div>
   
</div>
`;
